File: battle/attack/AttackVolcano.java
package common.battle.attack;

import common.battle.entity.AbEntity;
import common.battle.entity.Entity;
import common.util.Data;

import java.util.ArrayList;
import java.util.List;

public class AttackVolcano extends AttackAb {
	public ContVolcano handler;
	protected boolean attacked = false;
	protected final List<Entity> vcapt = new ArrayList<>();

	private byte volcTime = VOLC_ITV;

	public AttackVolcano(Entity e, AttackAb a, double sta, double end, int vt) {
		super(e, a, sta, end, false);
		isCounter = a.isCounter;
		this.sta = sta;
		this.end = end;
		this.waveType = vt;
	}

	@Override
	public void capture() {
		List<AbEntity> le = model.b.inRange(touch, dire, sta, end, excludeLastEdge);

		capt.clear();

		for (AbEntity e : le)
			if (e instanceof Entity && !vcapt.contains((Entity) e)) {
				capt.add(e);
			}
	}

	@Override
	public void excuse() {
		process();

		if (volcTime == 0) {
			volcTime = VOLC_ITV;
			vcapt.clear();
		} else {
			volcTime--;
		}

		if (attacker != null) {
			if (attacker.status[P_STRONG][0] != 0)
				atk += atk * attacker.status[P_STRONG][0] / 100;
			if (attacker.status[P_WEAK][0] != 0)
				atk = atk * attacker.status[P_WEAK][1] / 100;
			if (attacker.dire == 1 && attacker.basis.canon.deco == DECO_BASE_WATER)
				atk *= attacker.basis.b.t().getDecorationMagnification(attacker.basis.canon.deco, Data.DECO_SURGE);
		}

		for (AbEntity e : capt) {
			if (e.isBase() && !(e instanceof Entity))
				continue;

			if (e instanceof Entity) {
				e.damaged(this);
				attacked = true;

				vcapt.add((Entity) e);
			}
		}
	}
}

File: battle/attack/AttackVolcano.kt
package common.battle.attack

import common.battle.entity.AbEntity
import common.battle.entity.Entity
import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.util.Data
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JL
import page.anim.AnimBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter
import java.util.*

class AttackVolcano(a: AttackSimple, sta: Double, end: Double) : AttackAb(a, sta, end) {
    var attacked = false
    protected val vcapt: HashMap<Entity, Int>
    override fun capture() {
        val le: List<AbEntity> = model.b.inRange(touch, dire, sta, end)
        capt.clear()
        for (e in le) if ((abi and Data.Companion.AB_ONLY == 0 || e.targetable(type)) && e is Entity && !vcapt.containsKey(e)) capt.add(e)
    }

    override fun excuse() {
        process()
        for (e in capt) {
            if (e.isBase()) continue
            if (e is Entity) {
                e.damaged(this)
                vcapt[e] = Data.Companion.VOLC_ITV
                attacked = true
            }
        }
        vcapt.entries.removeIf { ent: MutableMap.MutableEntry<Entity, Int> ->
            val n = ent.value - 1
            if (n > 0) ent.setValue(n)
            n == 0
        }
    }

    init {
        vcapt = HashMap()
        sta = sta
        end = end
        this.waveType = Data.Companion.WT_VOLC
    }
}
